File: Python/lc_162_find_peak_element.py
# Time: O(log(n))
# Space: O(1)
class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        l = 0
        r = len(nums)-1

        while l<r:
            mid = (l+r)//2

            if nums[mid]>nums[mid+1]:
                r=mid
            else:
                l=mid+1

        return l

File: Go/lc_162_find_peak_element.go
package Go

// O(logn)
func findPeakElement(nums []int) int {

	return binSearch(0, len(nums)-1, nums)
}

func binSearch(left, right int, nums []int) int {

	if left >= right {
		return left
	}

	mid := int((left + right) / 2)

	if nums[mid] > nums[mid+1] {
		return binSearch(left, mid, nums)
	}

	return binSearch(mid+1, right, nums)
}
